File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
# Shebang line specifying the interpreter to be used for executing the script

import rclpy # Importing the rclpy module for working with ROS 2 Python
from rclpy.node import Node # Importing the Node class from rclpy.node module
from example_interfaces.srv import AddTwoInts # Importing the AddTwoInts service message from the example_interfaces.srv module
from functools import partial

class AddTwoIntClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")  
               # Initializing the parent Node class with the node name "add_two_ints_client"
        self.call_add_two_ints_server(6,7)
        self.call_add_two_ints_server(6,78)
        self.call_add_two_ints_server(-6,7)
        
    def call_add_two_ints_server(self,a,b):
        self.client = self.create_client(AddTwoInts, "add_two_ints")         # Creating a client for the AddTwoInts service with the service name "add_two_ints"

        while not self.client.wait_for_service(1.0):             # Waiting for the service to become available
            self.get_logger().warning("Waiting for server Add Two Ints")

        request = AddTwoInts.Request()         # Creating a request object of type AddTwoInts.Request

        request.a =a 
        request.b = b         # Setting the values of 'a' and 'b' in the request object

        future = self.client.call_async(request)         # Calling the service asynchronously and getting a future object
        future.add_done_callback(partial(self.callback_call_add_two_ints,a=a,b=b))

    def callback_call_add_two_ints(self,future,a,b):
        try:
            response = future.result()             # Getting the response from the completed future
            self.get_logger().info(
                str(a) + " + " + str(b) + " = " + str(response.sum)             )
            # Logging the sum of 'a' and 'b' received in the response
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))             # Logging an error message if the service call fails


def main(args=None):
    rclpy.init(args=args)     # Initializing the ROS 2 Python client library

    node = AddTwoIntClientNode()     # Creating an instance of the AddTwoIntClientNode class

    rclpy.spin(node)     # Spinning the node

    rclpy.shutdown()     # Shutting down the ROS 2 Python client library


if __name__ == "__main__":
    main()                             # Calling the main function if the script is run directly

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"                       // Include the RCLCPP library
#include "example_interfaces/srv/add_two_ints.hpp" // Include the AddTwoInts service message

class AddTwoIntsNode : public rclcpp::Node
{
public:
    AddTwoIntsNode() : Node("add_two_ints")
    {
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints"); // Create a client for the AddTwoInts service
        while (!client->wait_for_service(std::chrono::seconds(1)))                              // Wait for the service to be available
        {
            if (!rclcpp::ok()) // Check if ROS 2 is still running
            {
                RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
            }
            RCLCPP_INFO(this->get_logger(), "Service not available, waiting again...");
        }

        for (int i = 1; i <= 5; i++) // Send five requests
        {
            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>(); // Create a request object
            request->a = 3 + i;                                                              // Set the request values
            request->b = 4 - 2 * i;

            auto future = client->async_send_request(request); // Send the request asynchronously

            // Process callbacks until the response is received or an error occurs
            while (rclcpp::ok() && rclcpp::spin_some(this) != rclcpp::executor::FutureReturnCode::SUCCESS)
            {
                // Wait for the response
            }

            if (future.is_ready()) // Check if the future has a response
            {
                auto response = future.get();                               // Get the response
                RCLCPP_INFO(this->get_logger(), "Sum: %ld", response->sum); // Print the result
            }
            else // if the request fails
            {
                RCLCPP_ERROR(this->get_logger(), "Failed to call service");
            }
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);                           // Initialize the ROS 2 communication
    auto node = std::make_shared<AddTwoIntsNode>();      // Create an instance of AddTwoIntsNode
    rclcpp::spin(node);
    rclcpp::shutdown();                                  // Shutdown the ROS 2 communication
    return 0;
}
